Source/EventFlow.MsSql/ReliablePublish/MsSqlReliablePublishPersistence.cs

```diff
@@ -0,0 +1,132 @@
+            {
+                AggregateId = aggregateIdentity.Value,
+                MinAggregateSequenceNumber = domainEvents.Min(x => x.AggregateSequenceNumber),
+                MaxAggregateSequenceNumber = domainEvents.Max(x => x.AggregateSequenceNumber),
+            };
+
+            await _msSqlConnection.ExecuteAsync(
+                    Label.Named("publishlog-commit"),
+                    CancellationToken.None, // Unable to Cancel
+                    @"INSERT INTO [dbo].[EventFlowPublishLog]
+                    (AggregateId, MinAggregateSequenceNumber, MaxAggregateSequenceNumber)
+                    VALUES
+                    (@AggregateId, @MinAggregateSequenceNumber, @MaxAggregateSequenceNumber)",
+                    item)
+                .ConfigureAwait(false);
+        }
+
+        public async Task<VerificationState> GetUnverifiedItemsAsync(int maxCount, CancellationToken cancellationToken)
+        {
+            var logItems = await _msSqlConnection.QueryAsync<PublishLogItem>(
+                    Label.Named("publishlog-select"),
+                    cancellationToken,
+                    "SELECT TOP(@Top) [Id], [AggregateId],[MinAggregateSequenceNumber], [MaxAggregateSequenceNumber] FROM [dbo].[EventFlowPublishLog] ORDER BY [Id]",
+                    new { Top = maxCount })
+                .ConfigureAwait(false);
+
+            var positions = await _msSqlConnection.QueryAsync<string>(
+                    Label.Named("publishlog-global-position-select"),
+                    cancellationToken,
+                    "SELECT TOP 1 [LastVerifiedPosition] FROM [dbo].[EventFlowPublishVerifyState]")
+                .ConfigureAwait(false);
+
+            return new VerificationState(
+                new GlobalPosition(positions.First()),
+                logItems);
+        }
+
+        public async Task MarkVerifiedAsync(
+            IReadOnlyCollection<IPublishVerificationItem> verifiedItems,
+            GlobalPosition newVerifiedPosition,
+            CancellationToken cancellationToken)
+        {
+            await _msSqlConnection.ExecuteAsync(
+                Label.Named("publishlog-global-position-update"),
+                cancellationToken,
+                "UPDATE [dbo].[EventFlowPublishVerifyState] SET LastVerifiedPosition = @LastVerifiedPosition",
+                new
+                {
+                    LastVerifiedPosition = newVerifiedPosition.Value
+                });
+
+            foreach (var publishVerificationItem in verifiedItems)
+            {
+                var logItem = (PublishLogItem)publishVerificationItem;
+
+                await _msSqlConnection.ExecuteAsync(
+                    Label.Named("publishlog-confirm"),
+                    cancellationToken,
+                    "DELETE FROM [dbo].[EventFlowPublishLog] WHERE Id = @Id",
+                    new { logItem.Id});
+            }
+        }
+
+        private sealed class PublishLogItem : IPublishVerificationItem
```

Comment:
This is a DTO/DBO like structure which should contain only properties. Please put the logic somewhere else (extension method or user of this DTO), so IsPublished and IsFinalEvent don't have to be implemented in every persistence provider.

Reply:
Information can be related to storage: for example you can store bit column in EventFlow table. If we agree that we will store some kind of batch/markers i will change this.

Reply:
Yes, you're right - now that I think about it, your solution is more generic. Please ignore the comment.